### PR TITLE
General override member(s) functionality

### LINKED ISSUE #359: General override/implement member/function/variable operation
A override-method/member operation like many Java plugins has (e.g, lsp-java for Emacs). Not just abstract/unimplemented ones, but making it possible to override methods/members in general. Some interfaces have default implementations now, so these will probably not show up as abstract.

A lot of the Implement Abstract Members Quick Fix code can probably be reused with some minor tweaks.

### PR DESCRIPTION
**Pull request by themkat**
## General info
Override any member that is currently not implemented in the current class. This includes open functions, functions with default implementations, open members, abstract stuff etc. The main difference here is that this is general override member functionality, NOT JUST abstract members like the quick fix does. This is super useful for cases where you need to override a default implementation (e.g toString, Thread run, lsp4j interface methods like in TextDocumentService etc.), NOT just the unimplemented abstract ones. For me it would be super useful for my day job using Kotlin 😄 Having to google api docs to find signatures are a pain in the ass just for doing simple stuff like this. 

Two small issues:
- Implicit Any/Object methods does not seem to work. I can't really figure out why. The methods from them show up if the class has at least one superclass. Unsure on how to solve this, there are a TODO in the code at a relevant place. 
- No support for empty class bodies (i.e, no brackets). Not sure how to solve that. Probably not super important either. 

I think these are very minor. Just having this functionality in place would at least make me more productive using the language server 😄 

The quick fix for abstract methods and this new functionality, share a lot of code. I've moved most of the code to the override-package, as that was the most logical place to put it. 

Fixes #359 

This is a new protocol extension, so each lsp-client will need to have an implementation to use it. I've only done one for Emacs so far, and speaking of...

## Emacs screenshots
Running lsp-kotlin-override-members on MyClass (notice that multiple selections are possible!):
image

When we click enter:
image
(yes, it is the same sample project I often use for screenshot. A boring Spring Boot demo app)

Yes, it works for external interfaces and open classes as well, don't worry 🙂 There is an example of JDK Thread in the tests (in this PR, not in Emacs)

Emacs client implementation (lsp-mode):
https://github.com/themkat/lsp-mode/blob/kotlin_overrides/clients/lsp-kotlin.el#L224

Will make a PR for it once this PR has been merged. Just in case something changes.

~Will see if I can find the time to make a VSCode version or not. If not, I will make an issue in the vscode-kotlin repo with enough details to get someone else started~ 🙂 Really want to work on either Type hierarchy or Javadoc for Kotlin SDK and the JDK instead 😛 

**EDIT:** Made a VSCode version, as I was bored after work yesterday: https://github.com/fwcd/vscode-kotlin/pull/104

### CODE REVIEW

File: server/src/main/kotlin/org/javacs/kt/overridemembers/OverrideMembers.kt [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+        }
+        .flatten()
+
+private fun ClassDescriptor.canBeExtended() = this.kind.isInterface ||
+    this.modality == Modality.ABSTRACT ||
+    this.modality == Modality.OPEN
+
+private fun FunctionDescriptor.canBeOverriden() = (Modality.ABSTRACT == this.modality || Modality.OPEN == this.modality) && Modality.FINAL != this.modality && this.visibility != DescriptorVisibilities.PRIVATE && this.visibility != DescriptorVisibilities.PROTECTED
+
+private fun PropertyDescriptor.canBeOverriden() = (Modality.ABSTRACT == this.modality || Modality.OPEN == this.modality) && Modality.FINAL != this.modality && this.visibility != DescriptorVisibilities.PRIVATE && this.visibility != DescriptorVisibilities.PROTECTED
```
Comment (fwcd) [inline server/src/main/kotlin/org/javacs/kt/overridemembers/OverrideMembers.kt:118]: I _think_ we can do

```suggestion
private fun MemberDescriptor.canBeOverriden() = (Modality.ABSTRACT == this.modality || Modality.OPEN == this.modality) && Modality.FINAL != this.modality && this.visibility != DescriptorVisibilities.PRIVATE && this.visibility != DescriptorVisibilities.PROTECTED
```

Haven't checked whether that compiles though. That would let us get rid of some code duplication here and there.

File: server/src/main/kotlin/org/javacs/kt/overridemembers/OverrideMembers.kt [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+                    .getMemberScope(superClassTypeArguments)
+                    .getContributedDescriptors()
+                    .filter { classMember ->
+                        (classMember is FunctionDescriptor &&
+                         classMember.canBeOverriden() &&
+                         !overridesDeclaration(kotlinClass, classMember)) ||
+                        (classMember is PropertyDescriptor &&
+                         classMember.canBeOverriden() &&
+                         !overridesDeclaration(kotlinClass, classMember))
+                    }
```
Comment (fwcd) [inline server/src/main/kotlin/org/javacs/kt/overridemembers/OverrideMembers.kt:98]: I was going to propose factoring the `canBeOverriden` and `overridesDeclaration`, but then I saw that these are different overloads. Hmm. If we can use `MemberDescriptor`, we should be able to at least do

```suggestion
                    .filter { classMember ->
                        classMember.canBeOverriden() &&
                        (classMember is FunctionDescriptor && !overridesDeclaration(kotlinClass, classMember)) ||
                        (classMember is PropertyDescriptor && !overridesDeclaration(kotlinClass, classMember))
                    }
```

File: server/src/main/kotlin/org/javacs/kt/overridemembers/OverrideMembers.kt [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+                }
+            } else {
+                false
+            }
+        }
+
+fun overridesDeclaration(kotlinClass: KtClass, descriptor: PropertyDescriptor): Boolean =
+        kotlinClass.declarations.any {
+            it.name == descriptor.name.asString() && it.hasModifier(KtTokens.OVERRIDE_KEYWORD)
+        }
```
Comment (fwcd) [inline server/src/main/kotlin/org/javacs/kt/overridemembers/OverrideMembers.kt:165]: If I am not mistaken, we can shorten this to

```suggestion
// Checks if the class overrides the given declaration
fun overridesDeclaration(kotlinClass: KtClass, descriptor: FunctionDescriptor): Boolean =
        kotlinClass.declarations.any {
            it.name == descriptor.name.asString()
            && it.hasModifier(KtTokens.OVERRIDE_KEYWORD)
            && ((it as? KtNamedFunction)?.let { parametersMatch(it, descriptor) } ?? true)
        }

fun overridesDeclaration(kotlinClass: KtClass, descriptor: PropertyDescriptor): Boolean =
        kotlinClass.declarations.any {
            it.name == descriptor.name.asString() && it.hasModifier(KtTokens.OVERRIDE_KEYWORD)
        }
```

Maybe we could even merge these overloads:

```suggestion
// Checks if the class overrides the given declaration
fun overridesDeclaration(kotlinClass: KtClass, descriptor: MemberDescriptor): Boolean =
    when (descriptor) {
        is FunctionDescriptor -> kotlinClass.declarations.any {
            it.name == descriptor.name.asString()
            && it.hasModifier(KtTokens.OVERRIDE_KEYWORD)
            && ((it as? KtNamedFunction)?.let { parametersMatch(it, descriptor) } ?? true)
        }
        is PropertyDescriptor -> kotlinClass.declarations.any {
            it.name == descriptor.name.asString() && it.hasModifier(KtTokens.OVERRIDE_KEYWORD)
        }
        else -> false
    }
```

That would get simplify the `.filter { classMember -> ... }` above at the expense of a bit of type-safety, so I am not sure whether that would actually be more elegant.

File: server/src/main/kotlin/org/javacs/kt/overridemembers/OverrideMembers.kt [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+
+// Checks if two functions have matching parameters
+private fun parametersMatch(
+        function: KtNamedFunction,
+        functionDescriptor: FunctionDescriptor
+): Boolean {
```
Comment (fwcd) [inline server/src/main/kotlin/org/javacs/kt/overridemembers/OverrideMembers.kt:171]: Perhaps we could rewrite this into a big functional `zip`? I haven't checked on the specifics though, maybe there's too much specialized logic to do that.

**Review by fwcd — COMMENTED** [review]
Looks good, a few style suggestions here and there, but nothing too major 👍

File: server/src/main/kotlin/org/javacs/kt/overridemembers/OverrideMembers.kt [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+
+fun listOverridableMembers(file: CompiledFile, cursor: Int): List<CodeAction> {
+    val kotlinClass = file.parseAtPoint(cursor)
+
+    if (kotlinClass is KtClass) {
+        return createOverrideAlternatives(file, kotlinClass)
+    }
+
+    return emptyList()
+}
```
[the remark this answers is not in the document]
Comment (themkat) [reply server/src/main/kotlin/org/javacs/kt/overridemembers/OverrideMembers.kt:46]: We could in theory rewrite this to use a when-expression like above, but we need the `kotlinClass`-reference for the `createOverrideAlternatives`-call. That means the suggestion above won't compile due to unknown reference to `kotlinClass`. If we still need that variable, I don't think it improves much just using a when-expression instead of the if-version. Feel free to disagree 🙂

File: server/src/main/kotlin/org/javacs/kt/overridemembers/OverrideMembers.kt (hunk 2 above)
Comment (themkat) [reply server/src/main/kotlin/org/javacs/kt/overridemembers/OverrideMembers.kt:98]: Sadly no 🙁 Why? Because at this point `classMember` is sent in as a `DeclarationDescriptor` (`.getContributedDescriptors()` returns a `Collection<DeclarationDescriptor>`), so we would have to use another smart cast to make the above work.

File: server/src/main/kotlin/org/javacs/kt/overridemembers/OverrideMembers.kt (hunk 3 above)
Comment (themkat) [reply server/src/main/kotlin/org/javacs/kt/overridemembers/OverrideMembers.kt]: It prettifies a lot by making the code more compact, so I did that. See my comment above on the `filter`-expression and what the input is. We would probably have to smart cast to MemberDescriptor for the simplification to work. Unless I'm missing something. Will take another look tomorrow morning, as I'm starting to get tired now.

File: server/src/main/kotlin/org/javacs/kt/overridemembers/OverrideMembers.kt [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+import org.jetbrains.kotlin.psi.psiUtil.isAbstract
+import org.jetbrains.kotlin.psi.psiUtil.startOffset
+import org.jetbrains.kotlin.psi.psiUtil.unwrapNullability
+import org.jetbrains.kotlin.types.TypeProjection
+import org.jetbrains.kotlin.types.KotlinType
+import org.jetbrains.kotlin.lexer.KtTokens
+import org.jetbrains.kotlin.types.typeUtil.asTypeProjection
+
+// TODO: see where this should ideally be placed
+private const val DEFAULT_TAB_SIZE = 4
```
[the remark this answers is not in the document]
Comment (themkat) [reply server/src/main/kotlin/org/javacs/kt/overridemembers/OverrideMembers.kt:37]: Hmm. That's a great point. I can't find the exact issue, but I suggest we do this as part of another PR. Like you said, that is a bit out of scope for this PR... I prefer to keep the PRs as to-the-point as possible, which helps me keep focus on each task. 🙂 Please link the issue if you find it 🙂

File: server/src/main/kotlin/org/javacs/kt/overridemembers/OverrideMembers.kt (hunk 3 above)
Comment (themkat) [reply server/src/main/kotlin/org/javacs/kt/overridemembers/OverrideMembers.kt]: I was missing something 🙂 Took another look after getting some sleep, and found that just casting to `MemberDescriptor` in the filter-expression makes the expression more compact and expressive.  Done in [596c6a3](https://github.com/fwcd/kotlin-language-server/pull/379/commits/596c6a3268e2fa6d9cbeb13daa551e03af464e54).

File: server/src/main/kotlin/org/javacs/kt/overridemembers/OverrideMembers.kt (hunk 2 above)
Comment (themkat) [reply server/src/main/kotlin/org/javacs/kt/overridemembers/OverrideMembers.kt]: I think after the improvement in [596c6a3](https://github.com/fwcd/kotlin-language-server/pull/379/commits/596c6a3268e2fa6d9cbeb13daa551e03af464e54), there is not much more to be done. I think just casting to `MemberDescriptor` and calling the two methods in one go improves a lot 🙂

File: server/src/main/kotlin/org/javacs/kt/overridemembers/OverrideMembers.kt (hunk 4 above)
Comment (themkat) [reply server/src/main/kotlin/org/javacs/kt/overridemembers/OverrideMembers.kt:161]: A bit unsure about this. We loop through two different lists; value parameters and type parameters. Unsure on how much it would improve in this case, but I may miss something...

**Comment by themkat** [comment]
@fwcd , do you have time to look at this again soon? Would be good to get it done while the details are still semi-fresh in my memory. Sorry for nagging, know you are probably busy with other things 🙁

**Comment by themkat** [comment]
@fwcd , bump?

**Comment by fwcd** [comment]
Lgtm, sorry for the delay. Thanks!

**Comment by themkat** [comment]
@fwcd , no problem 🙂 Sorry for nagging you. Just very eager about this project, so I can get a little impatient 🙁
